### ISSUE TITLE
Steam: Ensure leaderboard submissions are more robust

### ISSUE THREAD

**Issue by ChrisNZL** [ChrisNZL/Tallowmere2#469]
Report from 0.1.2: local high score is saved locally, but has not appeared in Steam Leaderboards.

To do:
- Add on-screen confirmation if a new score is submitted to Steam's leaderboards successfully.
- Add messages/logs if scores fail to submit to Steam.
- Gracefully handle retries for Steam leaderboard submissions.

---

Related feedback ID: 20200521-LC938 (v0.1.3c)

```
5:31:41, Frame 26826, ERROR »  STEAM ERROR: LeaderboardEntrySubmitter_Steam failed to find or create leaderboard for leaderboard ID: XXXXXXXXX
>>>>> CRITICAL ERROR >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

GameStates: RompingThroughDungeon
GameSetupMode: Singleplayer
Players: 1
DungeonRoomTransitionerState: Null
Room: 10 / 10
RoomModifiers: None
YouHaveDiedState: Null

5:31:47, Frame 27159, WARNING »  WARNING: STEAM: OnStatsStored failed: k_EResultNoConnection

5:31:48, Frame 27249, WARNING »  GameClientFeedbackSubmitter: webRequest.error:

Request timeout

5:32:47, Frame 30759, WARNING »  WARNING: STEAM: OnStatsStored failed: k_EResultNoConnection
```

**Comment by ChrisNZL** [comment]
Additionally, just have scores be submitted each time a new room is reached? Don't have to rely on death happening?

**Comment by ChrisNZL** [comment]
v0.2.0 – auto bug reports – related Feedback IDs:

- 20201205-N6XWC

```
4:43:46, Frame 39234, ERROR »  STEAM ERROR: LeaderboardEntrySubmitter_Steam failed to find or create leaderboard for leaderboard ID: v:0.2 gm:1 gsm:1 p:1 am:1
>>>>> CRITICAL ERROR >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

GameStates: RompingThroughDungeon
GameSetupMode: Singleplayer
Players: 1
DungeonRoomTransitionerState: Null
Room: 1 / 1
RoomModifiers: None
YouHaveDiedState: Null
SystemPlayer InputDevice: 键盘
HumanPlayer 1 InputDevice: 键盘
```

**Comment by ChrisNZL** [comment]
Fixed in 0.2.2.

Leaderboard submissions are now cached/queued in a file, and only removed from the file upon successful submission.

If the player is offline, or can't access Steam's leaderboard system for whatever reason, the game client will attempt to submit every 5-60 seconds.

If leaderboards can't be found due to Internet access issues or request timeouts, the game will try submitting scores the next time the game is launched.
